bmc-acf/acf_manager.cpp

```diff
@@ -0,0 +1,461 @@
+        "org.freedesktop.DBus.Properties", "Get");
+    properties.append(inf);
+    properties.append(prop);
+    try
+    {
+        auto result = bus.call(properties);
+        if (!result.is_method_error())
+        {
+            std::variant<std::string> val;
+            result.read(val);
+            if (auto pVal = std::get_if<std::string>(&val))
+            {
+                propSerialNum.assign((pVal->data()), pVal->size());
+            }
+            else
+            {
+                log<level::ERR>("could not get the host's serial number\n");
+            }
+        }
+    }
+    catch (const std::exception& exc)
+    {
+        log<level::ERR>("dbus call for getting serial number failured: ");
+        propSerialNum = "";
+    }
+    return propSerialNum;
+}
+
+int readFieldModeProperty(const std::string& obj, const std::string& inf,
+                          const std::string& prop)
+{
+    bool propBool = false;
+    std::string object = obj;
+    auto bus = sdbusplus::bus::new_default();
+    auto properties = bus.new_method_call(
+        "xyz.openbmc_project.Software.BMC.Updater", object.c_str(),
+        "org.freedesktop.DBus.Properties", "Get");
+    properties.append(inf);
+    properties.append(prop);
+    try
+    {
+        auto result = bus.call(properties);
+        if (!result.is_method_error())
+        {
+            std::variant<bool> val{false};
+            result.read(val);
+            if (auto pVal = std::get_if<bool>(&val))
+            {
+                log<level::ERR>(
+                    "dbus call for getting FieldModeProperty failured: ");
+                propBool = (*pVal);
+            }
+        }
+    }
+    catch (const std::exception& exc)
+    {
+        log<level::ERR>("dbus call failure ");
+        return FAILURE;
+    }
+
+    return (int)propBool;
+}
+
+CeLogin::CeLoginRc
+    verifyAcfSerialNumberAndExpiration(std::vector<uint8_t>& accessControlFile,
+                                       std::vector<uint8_t>& publicKeyFile,
+                                       std::string& sDate)
+{
+    CeLogin::CELoginSequenceV1* sDecodedAsn = NULL;
+    CeLogin::CeLoginRc sRc = CeLogin::CeLoginRc::Failure;
+    CeLogin::CeLoginJsonData* sJsonData = NULL;
+
+    sRc = decodeAndVerifySignature(
+        accessControlFile.data(), accessControlFile.size(),
+        publicKeyFile.data(), publicKeyFile.size(), sDecodedAsn);
+
+    if (CeLogin::CeLoginRc::Success == sRc)
+    {
+        log<level::INFO>("Signature verified");
+
+        std::string mSerialNumber("");
+        mSerialNumber = readMachineSerialNumberProperty(
+            DBUS_INVENTORY_SYSTEM_OBJECT, DBUS_INVENTORY_ASSET_INTERFACE,
+            DBUS_SERIAL_NUM_PROP);
+
+        // If serial number is empty on machine set as UNSET for check with acf
+        if (mSerialNumber.empty())
+        {
+            mSerialNumber = UNSET_SERIAL_NUM_KEYWORD;
+        }
+        sJsonData = (CeLogin::CeLoginJsonData*)OPENSSL_malloc(
```

Comment:
memory getting freed in the failing path and regular path?

Reply:
auto sJsonData =
    std::unique_ptr<CeLogin::CeLoginJsonData, decltype(OPENSSL_free)*>{
        reinterpret_cast<CeLogin::CeLoginJsonData*>(malloc(sizeof(CeLogin::CeLoginJsonData))),
        OPENSSL_free };

Check if this can be done that way we need not worry about deleting the pointer

Reply:
I had errors trying to use the smart_pointer. I just left it as is.